packages/astro/src/content/mutable-data-store.ts

```diff
@@ -485,3 +485,162 @@ export interface MetaStore {
 	has: (key: string) => boolean;
 	delete: (key: string) => void;
 }
+
+export class InMemoryDataStore implements DataStore {
+	#collection = new Map<string, any>();
+	#events: Array<
+		| {
+				type: 'set';
+				entry: DataEntry;
+		  }
+		| {
+				type: 'delete';
+				key: string;
+		  }
+		| {
+				type: 'clear';
+		  }
+		| {
+				type: 'addAssetImport';
+				assetImport: string;
+				filePath: string;
+		  }
+		| {
+				type: 'addAssetImports';
+				assets: Array<string>;
+				filePath: string;
+		  }
+		| {
+				type: 'addModuleImport';
+				filePath: string;
+		  }
+	> = [];
+
+	get events() {
+		return this.#events;
+	}
+
+	get<TData extends Record<string, unknown> = Record<string, unknown>>(
+		key: string,
+	): DataEntry<TData> {
+		return this.#collection.get(key);
+	}
+
+	entries() {
+		return [...this.#collection.entries()];
+	}
+
+	values() {
+		return [...this.#collection.values()];
+	}
+
+	keys() {
+		return [...this.#collection.keys()];
+	}
+
+	set<TData extends Record<string, unknown>>(inputEntry: DataEntry<TData>) {
+		this.#events.push({
+			type: 'set',
+			entry: inputEntry,
+		});
+		const {
+			id: key,
+			data,
+			body,
+			filePath,
+			deferredRender,
+			digest,
+			rendered,
+			assetImports,
+		} = inputEntry;
+		if (!key) {
+			throw new Error(`ID must be a non-empty string`);
+		}
+		const id = String(key);
+		if (digest) {
+			const existing = this.#collection.get(id) as DataEntry | undefined;
+			if (existing && existing.digest === digest) {
+				return false;
+			}
+		}
+		const foundAssets = new Set<string>(assetImports);
+		// Check for image imports in the data. These will have been prefixed during schema parsing
+		new Traverse(data).forEach((_, val) => {
+			if (typeof val === 'string' && val.startsWith(IMAGE_IMPORT_PREFIX)) {
+				const src = val.replace(IMAGE_IMPORT_PREFIX, '');
+				foundAssets.add(src);
+			}
+		});
```

Comment:
If we're doing this two-pass approach, could we find a better way to handle images? This has always felt like a hack, but was needed because of the time when it had to run